src/wazuh_modules/wm_database.c

```diff
@@ -250,6 +250,38 @@ void wm_check_agents() {
     }
 }
 
+int wm_truncate_groups_file(const char *id, char *groups) {
+    char *temp_groups;
+    os_strdup(groups, temp_groups);
+    char *group;
+    int groups_number = 0;
+    int multigroup_length = 0;
+    bool groups_number_exceeded = true;
+
+    for(groups_number = 0; groups_number < MAX_GROUPS_PER_MULTIGROUP; ++groups_number) {
```

Comment:
strlen and strrchr (which is implemented in base of [strchr](https://github.com/bminor/glibc/blob/master/string/strrchr.c) ) will scan the whole string on each pass which could be 'slow' on larger strings (linear complexity)

You can avoid that and also the extra allocation by doing the whole operation in place and just reading the string backwards doing something like:

```
#define END_MARKER -128
static int wm_truncate_groups_file_no_copy(char *groups) {
    if(groups == 0 || groups[0] == 0){
        //file has no contents
        return -1;
    }

    char first_char = groups[0];
    groups[0] = END_MARKER;
    char* groups_end = groups + strlen(groups); //The len could be passed in as we know it from reading the file

    char* groups_str = groups_end;
    int group_count = 0;
    while(group_count < MAX_GROUPS_PER_MULTIGROUP) {

        if(groups_str[0] == END_MARKER){
            groups[0] = first_char;
            break;
        }

        group_count += (groups_str[0] == MULTIGROUP_SEPARATOR);

        groups_str--;
    }

    if(group_count >= MAX_GROUPS_PER_MULTIGROUP){
        memmove(groups, groups_str + 2 /*we stopped before a ','*/, groups_end - groups_str);
    }

    return 0;
}
```

You can see a quick benchmark comparison of this in https://quick-bench.com/q/OxnWGrOV0h5-HOTDu7FhwIZFjOI (obviously the speed difference is less noticeable with shorter strings)